### ISSUE TITLE
ray.wait hangs on node failure

### ISSUE THREAD

**Issue by richardliaw** [ray-project/ray#3264]
### Describe the problem

If you remove a node where `ray.wait` is waiting on a result, wait will never return. Is this intended?

```python
import pytest
import ray
import time


@pytest.fixture
def start_connected_cluster():
    # Start the Ray processes.
    g = Cluster(
        initialize_head=True,
        connect=True,
        head_node_args={
            "resources": dict(CPU=1),
            #"_internal_config": json.dumps({
            #    "num_heartbeats_timeout": 10
            #}
            )
        })
    yield g
    # The code after the yield will run as teardown code.
    ray.shutdown()
    g.shutdown()


def test_actor_failure(start_connected_cluster):
    cluster = start_connected_cluster
    node = cluster.add_node(resources=dict(CPU=1))

    @ray.remote(num_cpus=1)
    class TestActor(object):
        def sleeps(self):
            time.sleep(10)
            return 123

    two_actors = [TestActor.remote() for i in range(2)]
    two_actors_ret = [act.sleeps.remote() for act in two_actors]
    cluster.remove_node(node)
    cluster.wait_for_nodes()
    assert ray.global_state.cluster_resources()["CPU"] == 1

    start = time.time()
    res, _ = ray.wait(two_actors_ret, num_returns=2, timeout=20000)
    duration = time.time() - start
    assert duration < 20
```

cc: @robertnishihara @stephanie-wang

**Comment by richardliaw** [comment]
Another test case is the following.

`pytest.timeout` seems to be the canonical way of writing tests that have a time limit. 

Unfortunately, the following code doesn't respect the timeout:

```python
import pytest
import ray

# Test should finish running in 10 seconds
@pytest.mark.timeout(10)
def test_unfortunate():
    ray.init()

    def test():
        # sleeps for 30 seconds
        import time
        time.sleep(30)
        return 132

    test_remote = ray.remote(test)
    broken_promise = test_remote.remote()
    # unfortunately hangs here for 30 seconds...
    res, _ = ray.wait([broken_promise])
    assert res == 132
```

The `timeout` impl isn't completely esoteric either...
from `https://bitbucket.org/pytest-dev/pytest-timeout/src/fe7aed25207c1bc37cd5b76a799dc4738be44fb6/pytest_timeout.py`

```python

def timeout_setup(item):
    """Setup up a timeout trigger and handler"""
    params = get_params(item)
    if params.timeout is None or params.timeout <= 0:
        return
    if params.method == 'signal':
        def handler(signum, frame):
            __tracebackhide__ = True
            timeout_sigalrm(item, params.timeout)
        def cancel():
            signal.setitimer(signal.ITIMER_REAL, 0)
            signal.signal(signal.SIGALRM, signal.SIG_DFL)
        item.cancel_timeout = cancel
        signal.signal(signal.SIGALRM, handler)
        signal.setitimer(signal.ITIMER_REAL, params.timeout)
    elif params.method == 'thread':
        timer = threading.Timer(params.timeout, timeout_timer,
                                (item, params.timeout))
        def cancel():
            timer.cancel()
            timer.join()
        item.cancel_timeout = cancel
        timer.start()
```

**Comment by richardliaw** [comment]
This test case works now.

**Comment by soundway** [comment]
@richardliaw Sorry to comment on an old issue, but I'm curious if there's plan to make it possible for signal based timeout to work when ray.wait() is being called?

**Comment by richardliaw** [comment]
Hmm, can you make a new issue?

Sounds like a thing worth investing in :)
